From 4801820ba947ca895721ac0a198362409cd94d69 Mon Sep 17 00:00:00 2001
From: Thomas Guillem <thomas@gllm.fr>
Date: Tue, 30 Jul 2019 18:02:14 +0200
Subject: [PATCH 3/4] Fix getlogin() usage

Use the reentrant version (the getlogin() string was statically allocated and
could be overwritten on subsequent calls).

Also check for error and use "Guest" as a fallback.
---
 lib/init.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/lib/init.c b/lib/init.c
index eab69a5..3c01774 100644
--- a/lib/init.c
+++ b/lib/init.c
@@ -52,7 +52,7 @@
 #define MAX_URL_SIZE 256
 
 #ifdef _MSC_VER
-#define getlogin() "Guest"
+#define getlogin_r() ENXIO
 #define random rand
 #define getpid GetCurrentProcessId
 #endif // _MSC_VER
@@ -60,7 +60,7 @@
 #ifdef ESP_PLATFORM
 #include <esp_system.h>
 #define random esp_random
-#define getlogin() "Guest"
+#define getlogin_r() ENXIO
 #endif
 
 static int
@@ -206,7 +206,8 @@ void smb2_destroy_url(struct smb2_url *url)
 struct smb2_context *smb2_init_context(void)
 {
         struct smb2_context *smb2;
-        int i;
+        char buf[1024];
+        int i, ret;
 
         smb2 = malloc(sizeof(struct smb2_context));
         if (smb2 == NULL) {
@@ -214,7 +215,8 @@ struct smb2_context *smb2_init_context(void)
         }
         memset(smb2, 0, sizeof(struct smb2_context));
 
-        smb2_set_user(smb2, getlogin());
+        ret = getlogin_r(buf, sizeof(buf));
+        smb2_set_user(smb2, ret == 0 ? buf : "Guest");
         smb2->fd = -1;
         smb2->sec = SMB2_SEC_UNDEFINED;
         smb2->version = SMB2_VERSION_ANY;
-- 
2.20.1

